File: Data Structure & Algorithm/Template/KMP/KMP.ts
export function KMPSearch(text: string, pattern: string) {
  if (pattern.length === 0) {
    return 0;
  }

  // find Long String Prefix
  const findLSP = (pattern: string): number[] => {
    const lsp = new Array(pattern.length).fill(0);
    let j = 0; // j is the tail of prefix
    lsp[0] = 0;

    // i is the tail of suffix
    for (let i = 1; i < pattern.length; i++) {
      while (j > 0 && pattern[i] !== pattern[j]) {
        j = lsp[j - 1];
      }
      if (pattern[i] === pattern[j]) {
        j++;
      }
      lsp[i] = j;
    }

    return lsp;
  };

  const lsp = findLSP(pattern);
  let j = 0;

  for (let i = 0; i < text.length; i++) {
    while (j > 0 && text[i] !== pattern[j]) {
      j = lsp[j - 1];
    }
    if (text[i] === pattern[j]) {
      if (j === pattern.length - 1) {
        return i - j;
      }
      j++;
    }
  }

  return -1;
}

// test
const res = KMPSearch('AABABABABC', 'ABABC');
console.log(res);

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Algorithm/KMP/KMP.py
from typing import List


def KMPSearch(text: str, pattern: str) -> int:
    def findLSP(pattern: str) -> List[int]:
        lsp = [0] * len(pattern)
        length = 0
        i = 1

        while i < len(pattern):
            if pattern[i] == pattern[length]:
                length = length + 1
                lsp[i] = length
                i = i + 1
            else:
                if length != 0:
                    length = lsp[length-1]
                else:
                    lsp[i] = 0
                    i = i + 1

        return lsp

    i = 0
    j = 0
    lsp = findLSP(pattern)

    while i < len(text):
        if text[i] == pattern[j]:
            i = i + 1
            j = j + 1
        elif j > 0:
            j = lsp[j-1]
        else:
            i = i + 1

        if j == len(pattern):
            return i - j
    return -1


# test
text = "AABABABABC"
pattern = "ABABC"
res = KMPSearch(text, pattern)
print(res)
